cmd/k8s-operator/connector.go

```diff
@@ -0,0 +1,262 @@
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get tailscale.io Connector: %w", err)
+	}
+	if !cn.DeletionTimestamp.IsZero() {
+		logger.Debugf("connector is being deleted or should not be exposed, cleaning up components")
+		ix := xslices.Index(cn.Finalizers, FinalizerName)
+		if ix < 0 {
+			logger.Debugf("no finalizer, nothing to do")
+			return reconcile.Result{}, nil
+		}
+
+		if done, err := a.maybeCleanupSubnetRouter(ctx, logger, cn); err != nil {
+			return reconcile.Result{}, err
+		} else if !done {
+			logger.Debugf("cleanup not finished, will retry...")
+			return reconcile.Result{RequeueAfter: shortRequeue}, nil
+		}
+
+		cn.Finalizers = append(cn.Finalizers[:ix], cn.Finalizers[ix+1:]...)
+		if err := a.Update(ctx, cn); err != nil {
+			return reconcile.Result{}, err
+		}
+		logger.Infof("connector resources cleaned up")
+		return reconcile.Result{}, nil
+	}
+
+	oldCnStatus := cn.Status.DeepCopy()
+	defer func() {
+		if cn.Status.SubnetRouter == nil {
+			tsoperator.SetConnectorCondition(cn, tsapi.ConnectorReady, metav1.ConditionUnknown, "", "", cn.Generation, a.clock, logger)
+		} else if cn.Status.SubnetRouter.Ready == metav1.ConditionTrue {
+			tsoperator.SetConnectorCondition(cn, tsapi.ConnectorReady, metav1.ConditionTrue, reasonSubnetRouterCreated, reasonSubnetRouterCreated, cn.Generation, a.clock, logger)
+		} else {
+			tsoperator.SetConnectorCondition(cn, tsapi.ConnectorReady, metav1.ConditionFalse, cn.Status.SubnetRouter.Reason, cn.Status.SubnetRouter.Reason, cn.Generation, a.clock, logger)
+		}
+		if !apiequality.Semantic.DeepEqual(oldCnStatus, cn.Status) {
+			// an error encountered here should get returned by the Reconcile function
+			if updateErr := a.Client.Status().Update(ctx, cn); updateErr != nil {
+				err = updateErr
+			}
+		}
+	}()
+
+	if !slices.Contains(cn.Finalizers, FinalizerName) {
+		// This log line is printed exactly once during initial provisioning,
+		// because once the finalizer is in place this block gets skipped. So,
+		// this is a nice place to tell the operator that the high level,
+		// multi-reconcile operation is underway.
+		logger.Infof("ensuring connector is set up")
+		cn.Finalizers = append(cn.Finalizers, FinalizerName)
+		if err := a.Update(ctx, cn); err != nil {
+			err = fmt.Errorf("failed to add finalizer: %w", err)
+			logger.Errorf("error adding finalizer: %v", err)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// At the momement SubnetRouter is the only component configurable via
+	// ConnectorSpec and a ConnectorSpec without a SubnetRouter is invalid- but
+	// that will change in the future - so run the cleanup and provision
+	// conditionally.
+	if cn.Spec.SubnetRouter != nil && len(cn.Spec.SubnetRouter.Routes) > 0 {
+		var sb strings.Builder
+		sb.WriteString(string(cn.Spec.SubnetRouter.Routes[0]))
+		for _, r := range cn.Spec.SubnetRouter.Routes[1:] {
```

Comment:
control (or tailscaled, i forget which) rejects routes that have `pfx.Masked() != pfx`, we might want to check that here.

Reply:
https://pkg.go.dev/net/netip#Prefix.Masked

Reply:
Would be nice to eventually be able to put this into the validation at apply time, but CEL does not have IP parsing, not sure if this would be too expensive for them to ever do it.